README.md
[![NPM version][npm-version-image]][npm-url] [![NPM downloads][npm-downloads-image]][npm-url] [![MIT License][license-image]][license-url] [![Build Status][travis-image]][travis-url]

A tiny javascript library to count in natural language e.g. Zero, One, Two, Three.. up to 999,999,999,999,999.

## Documentation

### Basic Usage

Here's all this does so far:

    // Factory's name will be "FactoryFive"
    _n = 'Factory' + count(5).camel();
    
    // Unit tests
    test.equal(count(764).camel(), '"SevenHundredSixtyFour"');    
    test.equal(count(999999999999999).camel(), 'NineHundredNinetyNineTrillionNineHundredNinetyNineBillionNineHundredNinetyNineMillionNineHundredNinetyNineThousandNineHundredNinetyNine');
   
## [Changelog](CHANGELOG.md)

## [Contributing](CONTRIBUTING.md)

## License

Count.js is freely distributable under the terms of the [MIT license](LICENSE).

[license-image]: http://img.shields.io/badge/license-MIT-blue.svg?style=flat
[license-url]: LICENSE

[npm-url]: https://npmjs.org/package/count
[npm-version-image]: http://img.shields.io/npm/v/counts.svg?style=flat
[npm-downloads-image]: http://img.shields.io/npm/dm/counts.svg?style=flat

[travis-url]: http://travis-ci.org/nickckaye/count
[travis-image]: http://img.shields.io/travis/nickckaye/count/master.svg?style=flat


CHANGELOG.md
Changelog
=========

### 0.1.0

Initial release



test/count/camel.js
var count = require('../../count');

exports.isCount = {
  setUp: function (done) {
    done();
  },

  'CamelCase from 0 to 9': function (test) {
    test.equal(count(0).camel(), 'Zero');
    test.equal(count(1).camel(), 'One');
    test.equal(count(2).camel(), 'Two');
    test.equal(count(3).camel(), 'Three');
    test.equal(count(4).camel(), 'Four');
    test.equal(count(5).camel(), 'Five');
    test.equal(count(6).camel(), 'Six');
    test.equal(count(7).camel(), 'Seven');
    test.equal(count(8).camel(), 'Eight');
    test.equal(count(9).camel(), 'Nine');
    test.done();
  },

  'CamelCase from 10 to 19': function (test) {
    test.equal(count(10).camel(), 'Ten');
    test.equal(count(11).camel(), 'Eleven');
    test.equal(count(12).camel(), 'Twelve');
    test.equal(count(13).camel(), 'Thirteen');
    test.equal(count(14).camel(), 'Fourteen');
    test.equal(count(15).camel(), 'Fifteen');
    test.equal(count(16).camel(), 'Sixteen');
    test.equal(count(17).camel(), 'Seventeen');
    test.equal(count(18).camel(), 'Eighteen');
    test.equal(count(19).camel(), 'Nineteen');
    test.done();
  },

  'CamelCase from 20 to 99': function (test) {
    test.equal(count(20).camel(), 'Twenty');
    test.equal(count(21).camel(), 'TwentyOne');
    test.equal(count(22).camel(), 'TwentyTwo');
    test.equal(count(34).camel(), 'ThirtyFour');
    test.equal(count(42).camel(), 'FortyTwo');
    test.equal(count(57).camel(), 'FiftySeven');
    test.equal(count(64).camel(), 'SixtyFour');
    test.equal(count(75).camel(), 'SeventyFive');
    test.equal(count(89).camel(), 'EightyNine');
    test.done();
  },

  'CamelCase from 100 to 900': function (test) {
    test.equal(count(110).camel(), 'OneHundredTen');
    test.equal(count(221).camel(), 'TwoHundredTwentyOne');
    test.equal(count(322).camel(), 'ThreeHundredTwentyTwo');
    test.equal(count(434).camel(), 'FourHundredThirtyFour');
    test.equal(count(542).camel(), 'FiveHundredFortyTwo');
    test.equal(count(657).camel(), 'SixHundredFiftySeven');
    test.equal(count(764).camel(), 'SevenHundredSixtyFour');
    test.equal(count(875).camel(), 'EightHundredSeventyFive');
    test.equal(count(989).camel(), 'NineHundredEightyNine');
    test.done();
  },

  'CamelCase from 1000 to 999000': function (test) {
    test.equal(count(1110).camel(), 'OneThousandOneHundredTen');
    test.equal(count(15221).camel(), 'FifteenThousandTwoHundredTwentyOne');
    test.equal(count(75322).camel(), 'SeventyFiveThousandThreeHundredTwentyTwo');
    test.equal(count(285434).camel(), 'TwoHundredEightyFiveThousandFourHundredThirtyFour');
    test.equal(count(792542).camel(), 'SevenHundredNinetyTwoThousandFiveHundredFortyTwo');
    test.done();
  },

  'CamelCase from 1000000 to 999000000': function (test) {
    test.equal(count(1001110).camel(), 'OneMillionOneThousandOneHundredTen');
    test.equal(count(40015221).camel(), 'FortyMillionFifteenThousandTwoHundredTwentyOne');
    test.equal(count(60075322).camel(), 'SixtyMillionSeventyFiveThousandThreeHundredTwentyTwo');
    test.equal(count(700285434).camel(), 'SevenHundredMillionTwoHundredEightyFiveThousandFourHundredThirtyFour');
    test.equal(count(900792542).camel(), 'NineHundredMillionSevenHundredNinetyTwoThousandFiveHundredFortyTwo');
    test.done();
  },

  'CamelCase from 1000000000 to 999000000000': function (test) {
    test.equal(count(1001001110).camel(), 'OneBillionOneMillionOneThousandOneHundredTen');
    test.equal(count(2340015221).camel(), 'TwoBillionThreeHundredFortyMillionFifteenThousandTwoHundredTwentyOne');
    test.equal(count(7860075322).camel(), 'SevenBillionEightHundredSixtyMillionSeventyFiveThousandThreeHundredTwentyTwo');
    test.equal(count(432700285434).camel(), 'FourHundredThirtyTwoBillionSevenHundredMillionTwoHundredEightyFiveThousandFourHundredThirtyFour');
    test.equal(count(584900792542).camel(), 'FiveHundredEightyFourBillionNineHundredMillionSevenHundredNinetyTwoThousandFiveHundredFortyTwo');
    test.done();
  },

  'CamelCase from 1000000000000 to 999000000000000': function (test) {
    test.equal(count(1001001001110).camel(), 'OneTrillionOneBillionOneMillionOneThousandOneHundredTen');
    test.equal(count(2032340015221).camel(), 'TwoTrillionThirtyTwoBillionThreeHundredFortyMillionFifteenThousandTwoHundredTwentyOne');
    test.equal(count(3887860075322).camel(), 'ThreeTrillionEightHundredEightySevenBillionEightHundredSixtyMillionSeventyFiveThousandThreeHundredTwentyTwo');
    test.equal(count(654432700285434).camel(), 'SixHundredFiftyFourTrillionFourHundredThirtyTwoBillionSevenHundredMillionTwoHundredEightyFiveThousandFourHundredThirtyFour');
    test.equal(count(987584900792542).camel(), 'NineHundredEightySevenTrillionFiveHundredEightyFourBillionNineHundredMillionSevenHundredNinetyTwoThousandFiveHundredFortyTwo');
    test.equal(count(999999999999999).camel(), 'NineHundredNinetyNineTrillionNineHundredNinetyNineBillionNineHundredNinetyNineMillionNineHundredNinetyNineThousandNineHundredNinetyNine');
    test.done();
  }


};


test/count/words.js
var count = require('../../count');

exports.isCount = {
  setUp: function (done) {
    done();
  },

  'is an array of words': function (test) {
    // TODO: test Get the sequence of words that spells out the count count(*).words()
    test.done();
  }
};


test/count/is_count.js
var count = require('../../count');

exports.isCount = {
  setUp: function (done) {
    count.createFromInputFallback = function () {
      throw new Error('input not handled by count');
    };
    done();
  },

  'is count object': function (test) {
    test.expect(12);

    var MyObj = function () {
      },
      extend = function (a, b) {
        var i;
        for (i in b) {
          a[i] = b[i];
        }
        return a;
      };
    MyObj.prototype.toDate = function () {
      return new Date();
    };

    test.ok(count.isCount(count()), 'simple count object');
    test.ok(count.isCount(count(null)), 'invalid count object');
    test.ok(count.isCount(extend({}, count())), 'externally cloned moments are moments');

    test.ok(!count.isCount(new MyObj()), 'myObj is not count object');
    test.ok(!count.isCount(count), 'count function is not count object');
    test.ok(!count.isCount(new Date()), 'date object is not count object');
    test.ok(!count.isCount(Object), 'Object is not count object');
    test.ok(!count.isCount('foo'), 'string is not count object');
    test.ok(!count.isCount(1), 'number is not count object');
    test.ok(!count.isCount(NaN), 'NaN is not count object');
    test.ok(!count.isCount(null), 'null is not count object');
    test.ok(!count.isCount(undefined), 'undefined is not count object');

    test.done();
  },

  'is count with hacked hasOwnProperty': function (test) {
    var obj = {};
    // HACK to suppress jshint warning about bad property name
    obj['hasOwnMoney'.replace('Money', 'Property')] = function () {
      return true;
    };

    test.ok(!count.isCount(obj), 'isCount works even if passed object has a wrong hasOwnProperty implementation (ie8)');
    test.done();
  }
};


Gruntfile.js
module.exports = function (grunt) {
  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),
    concat: {
      tests: {
        src: [
          'test/browser-prefix.js.txt',
          'test/count/*.js',
          'test/browser-suffix.js.txt'
        ],
        dest: 'min/tests.js'
      }
    },
    env: {
      sauceLabs: (grunt.file.exists('.sauce-labs.creds') ?
        grunt.file.readJSON('.sauce-labs.creds') : {})
    },
    karma: {
      options: {
        frameworks: ['nodeunit'],
        files: [
          'min/tests.js',
          'test/browser.js'
        ],
        sauceLabs: {
          startConnect: true,
          testName: 'CountJS'
        },
        customLaunchers: {
          slChromeWinXp: {
            base: 'SauceLabs',
            browserName: 'chrome',
            platform: 'Windows XP'
          },
          slIe9Win7: {
            base: 'SauceLabs',
            browserName: 'internet explorer',
            platform: 'Windows 7',
            version: '9'
          },
          slIe8Win7: {
            base: 'SauceLabs',
            browserName: 'internet explorer',
            platform: 'Windows 7',
            version: '8'
          },
          slFfLinux: {
            base: 'SauceLabs',
            browserName: 'firefox',
            platform: 'Linux'
          },
          slSafariOsx: {
            base: 'SauceLabs',
            browserName: 'safari',
            platform: 'OS X 10.8'
          }
        }
      },
      server: {
        browsers: []
      },
      chrome: {
        singleRun: true,
        browsers: ['Chrome']
      },
      firefox: {
        singleRun: true,
        browsers: ['Firefox']
      },
      sauce: {
        options: {reporters: ['dots']},
        singleRun: true,
        browsers: [
          'slChromeWinXp',
          'slIe9Win7',
          'slIe8Win7',
          'slFfLinux',
          'slSafariOsx'
        ]
      }
    },

    uglify: {
      main: {
        files: {
          'min/count.min.js': 'count.js'
        }
      },
      options: {
        mangle: true,
        compress: {
          dead_code: false // jshint ignore:line
        },
        output: {
          ascii_only: true // jshint ignore:line
        },
        report: 'min',
        preserveComments: 'some'
      }
    },
    nodeunit: {
      all: ['test/count/**/*.js']
    },
    jshint: {
      all: [
        'Gruntfile.js', 'count.js', 'test/**/*.js',
        '!test/browser*.js'
      ],
      options: {
        'node': true,
        'browser': true,
        'boss': false,
        'curly': true,
        'debug': false,
        'devel': false,
        'eqeqeq': true,
        'eqnull': true,
        'evil': false,
        'forin': false,
        'immed': false,
        'laxbreak': false,
        'newcap': true,
        'noarg': true,
        'noempty': false,
        'nonew': false,
        'onevar': true,
        'plusplus': false,
        'regexp': false,
        'undef': true,
        'sub': true,
        'strict': false,
        'white': true,
        'es3': true,
        'camelcase': true,
        'globals': {
          'define': false
        }
      }
    },
    jscs: {
      all: [
        'Gruntfile.js', 'count.js',
        'test/**/*.js', '!test/browser*.js'
      ],
      options: {
        config: '.jscs.json'
      }
    },
    watch: {
      test: {
        files: [
          'count.js',
          'test/**/*.js'
        ],
        tasks: ['nodeunit']
      },
      jshint: {
        files: '<%= jshint.all %>',
        tasks: ['jshint']
      }
    },
    benchmark: {
      all: {
        src: ['benchmarks/*.js']
      }
    }
  });

  grunt.loadTasks('tasks');

  // These plugins provide necessary tasks.
  require('load-grunt-tasks')(grunt);

  // Default task.
  grunt.registerTask('default', ['jshint', 'jscs', 'nodeunit']);

  // test tasks
  grunt.registerTask('test', ['test:node', 'test:browser']);
  grunt.registerTask('test:node', ['nodeunit']);
  grunt.registerTask('test:server', ['concat', 'karma:server']);
  grunt.registerTask('test:browser', ['concat', 'karma:chrome', 'karma:firefox']);
  grunt.registerTask('test:sauce-browser', ['concat', 'env:sauceLabs', 'karma:sauce']);
  grunt.registerTask('test:travis-sauce-browser', ['concat', 'karma:sauce']);

  // travis build task
  grunt.registerTask('build:travis', [
    // code style
    'jshint', 'jscs',
    // node tests
    'test:node'
  ]);

  // Task to be run when releasing a new version
  grunt.registerTask('release', [
    'jshint', 'nodeunit', 'concat',
    'component', 'uglify:main'
  ]);
};


tasks/bump_version.js
module.exports = function (grunt) {
  grunt.registerTask('bump_version', function (version) {
    if (!version || version.split('.').length !== 3) {
      grunt.fail.fatal('malformed version. Use\n\n    grunt bump_version:1.2.3')
    }

    grunt.config('string-replace.count-js', {
      files: {'count.js': 'count.js'},
      options: {
        replacements: [
          {
            pattern: /\/\/! version : .*/,
            replacement: '//! version : ' + version
          },
          {
            pattern: /VERSION = '.*'/,
            replacement: "VERSION = '" + version + "'"
          }
        ]
      }
    });

    grunt.config('string-replace.package-json', {
      files: {'package.json': 'package.json'},
      options: {
        replacements: [
          {
            pattern: /"version": .*/,
            replacement: '"version": "' + version + '",'
          }
        ]
      }
    });

    grunt.config('string-replace.bower-json', {
      files: {'bower.json': 'bower.json'},
      options: {
        replacements: [
          {
            pattern: /"version": .*/,
            replacement: '"version": "' + version + '",'
          }
        ]
      }
    });

    grunt.config('string-replace.component-json', {
      files: {'component.json': 'component.json'},
      options: {
        replacements: [
          {
            pattern: /"version": .*/,
            replacement: '"version": "' + version + '",'
          }
        ]
      }
    });

    grunt.task.run([
      'string-replace:count-js',
      'string-replace:package-json',
      'string-replace:bower-json',
      'string-replace:component-json'
    ]);
  });
};


count.js
//! count.js
//! version : 0.1.3
//! authors : Nick Kaye
//! license : MIT
//! www.nickkaye.com/count-js

(function (undefined) {
  var count,
    VERSION = '0.1.3',
    words = {
      0: 'zero',
      1: 'one',
      2: 'two',
      3: 'three',
      4: 'four',
      5: 'five',
      6: 'six',
      7: 'seven',
      8: 'eight',
      9: 'nine',
      10: 'ten',
      11: 'eleven',
      12: 'twelve',
      13: 'thirteen',
      14: 'fourteen',
      15: 'fifteen',
      16: 'sixteen',
      17: 'seventeen',
      18: 'eighteen',
      19: 'nineteen',
      20: 'twenty',
      30: 'thirty',
      40: 'forty',
      50: 'fifty',
      60: 'sixty',
      70: 'seventy',
      80: 'eighty',
      90: 'ninety'
    },
    zeros = [
      {
        min: 2,
        max: 2,
        name: 'hundred'
      },
      {
        min: 3,
        max: 5,
        name: 'thousand'
      },
      {
        min: 6,
        max: 8,
        name: 'million'
      },
      {
        min: 9,
        max: 11,
        name: 'billion'
      },
      {
        min: 12,
        max: 14,
        name: 'trillion'
      }
    ],
    i,
    oldGlobalCount,
    hasOwnProperty = Object.prototype.hasOwnProperty,
    globalScope = typeof global !== 'undefined' ? global : this,
    hasModule = (typeof module !== 'undefined' && module.exports);

  /**
   * @param {Number} num
   * @returns {Array} words representing the number
   */
  function wordsForNumber(num) {
    var _n = num // number to work with
      , _p // power of ten
      , _m // minimum (calculator)
      , _i // iterator
      , _out = []; // output

    if (_n === 0) {
      _out.push(words[0]);
    } else {
      while (_n > 0) {
        //
        // from 1 to 19 e.g. "seventeen" (done)
        if (_n < 20) {
          _out.push(words[_n]);
          _n = 0;
        }
        // from 20 to 99 e.g. "eighty" (leave <10)
        else if (_n < 100) {
          _p = Math.floor(_n / 10);
          _out.push(words[_p * 10]);
          _n -= _p * 10;
        }
        // iterate through the possible x10 digits and reference zeros{} language dynamically
        else {
          for (_i in zeros) {
            if (_n < Math.pow(10, zeros[_i].max + 1)) {
              _m = Math.pow(10, zeros[_i].min);
              _p = Math.floor(_n / _m);
              _out = _out.concat(wordsForNumber(_p));
              _out.push(zeros[_i].name);
              _n -= _p * _m;
              break;
            }
          }
        }
      }
    }
    return _out;
  }

  function capitalize(str) {
    return str.charAt(0).toUpperCase() + str.slice(1);
  }

  function defaultParsingFlags() {
    // We need to deep clone this object,
    // and es5 standard is not very helpful.
    return {
      empty: false,
      nullInput: false,
      userInvalidated: false
    };
  }

  function printMsg(msg) {
    if (count.suppressDeprecationWarnings === false &&
      typeof console !== 'undefined' && console.warn) {
      console.warn('Deprecation warning: ' + msg);
    }
  }

  function deprecate(msg, fn) {
    var firstTime = true;
    return extend(function () {
      if (firstTime) {
        printMsg(msg);
        firstTime = false;
      }
      return fn.apply(this, arguments);
    }, fn);
  }

  /**
   * Constructor
   * @constructs Count
   * @param config
   */
  function Count(config) {
    copyConfig(this, config);
  }

  /**
   * Copy configuration to target object
   * @param to
   * @param from
   * @returns {*}
   */
  function copyConfig(to, from) {
    if (typeof from._isACountObject !== 'undefined') {
      to._isACountObject = from._isACountObject;
    }
    if (typeof from._i !== 'undefined') {
      to._i = from._i;
    }
    if (typeof from._pf !== 'undefined') {
      to._pf = from._pf;
    }
    return to;
  }

  /**
   * Simple object extension
   * @param a
   * @param b
   * @returns {*}
   */
  function extend(a, b) {
    for (var i in b) {
      if (hasOwnProp(b, i)) {
        a[i] = b[i];
      }
    }
    if (hasOwnProp(b, 'toString')) {
      a.toString = b.toString;
    }
    if (hasOwnProp(b, 'valueOf')) {
      a.valueOf = b.valueOf;
    }
    return a;
  }

  /**
   * Top-level: Make a Count object
   * @param config
   * @returns {*}
   */
  function makeCount(config) {
    var input = config._i;
    if (config._i === null) {
      return count.invalid({nullInput: true});
    }
    if (typeof input !== 'number') {
      config._i = Number(input);
    }
    return new Count(config);
  }

  /**
   * Top-level: Make this accessible
   * @param input
   * @returns {*}
   */
  count = function (input) {
    var c;

    // object construction must be done this way.
    // https://github.com/nickckaye/count/issues/1423
    c = {};
    c._isACountObject = true;
    c._i = input;
    c._pf = defaultParsingFlags();

    return makeCount(c);
  };

  count.suppressDeprecationWarnings = false;

  count.createFromInputFallback = deprecate(
      'count construction falls back to js Date. This is ' +
      'discouraged and will be removed in upcoming major ' +
      'release. Please refer to ' +
      'https://github.com/nickckaye/count/issues/1407 for more info.',
    function (config) {
      config._d = new Date(config._i);
    }
  );

  // version number
  count.version = VERSION;

  // compare count object
  count.isCount = function (obj) {
    return obj instanceof Count ||
      (obj != null && hasOwnProp(obj, '_isACountObject'));
  };

  count.invalid = function (flags) {
    var m = count(0);
    if (flags != null) {
      extend(m._pf, flags);
    } else {
      m._pf.userInvalidated = true;
    }
    return m;
  };

  /**
   * Count Prototype
   */

  extend(count.fn = Count.prototype, {

    /**
     * Create a camel case block out of the input.
     * @returns {string}
     */
    camel: function () {
      var out = '',
        words = this.words();
      for (i = 0; i < words.length; i++) {
        out += capitalize(words[i]);
      }
      return out;
    },

    /**
     * Get the sequence of words that spells out the count
     */
    words: function () {
      return wordsForNumber(this._i);
    },

    clone: function () {
      return count(this);
    },

    valueOf: function () {
      return +this._i;
    },

    parsingFlags: function () {
      return extend({}, this._pf);
    }

  });

  /**
   * Exposing Count
   */

  function hasOwnProp(a, b) {
    return hasOwnProperty.call(a, b);
  }

  function makeGlobal(shouldDeprecate) {
    /*global ender:false */
    if (typeof ender !== 'undefined') {
      return;
    }
    oldGlobalCount = globalScope.count;
    if (shouldDeprecate) {
      globalScope.count = deprecate(
          'Accessing Count through the global scope is ' +
          'deprecated, and will be removed in an upcoming ' +
          'release.',
        count);
    } else {
      globalScope.count = count;
    }
  }

  // CommonJS module is defined
  if (hasModule) {
    module.exports = count;
  } else if (typeof define === 'function' && define.amd) {
    define('count', function (require, exports, module) {
      if (module.config && module.config() && module.config().noGlobal === true) {
        // release the global variable
        globalScope.count = oldGlobalCount;
      }

      return count;
    });
    makeGlobal(true);
  } else {
    makeGlobal();
  }
}).call(this);
